internal/provider/cached_image_resource.go

```diff
@@ -293,128 +295,246 @@ func (r *CachedImageResource) Configure(ctx context.Context, req resource.Config
+		opts.RemoteRepoBuildMode = data.RemoteRepoBuildMode.ValueBool()
 	}
 
 	if !data.SSLCertBase64.IsNull() {
-		env["ENVBUILDER_SSL_CERT_BASE64"] = tfValueToString(data.SSLCertBase64)
+		overrides["ENVBUILDER_SSL_CERT_BASE64"] = struct{}{}
+		opts.SSLCertBase64 = data.SSLCertBase64.ValueString()
 	}
 
 	if !data.Verbose.IsNull() {
-		env["ENVBUILDER_VERBOSE"] = tfValueToString(data.Verbose)
+		overrides["ENVBUILDER_VERBOSE"] = struct{}{}
+		opts.Verbose = data.Verbose.ValueBool()
 	}
 
 	if !data.WorkspaceFolder.IsNull() {
-		env["ENVBUILDER_WORKSPACE_FOLDER"] = tfValueToString(data.WorkspaceFolder)
+		overrides["ENVBUILDER_WORKSPACE_FOLDER"] = struct{}{}
+		opts.WorkspaceFolder = data.WorkspaceFolder.ValueString()
 	}
 
-	// Do ExtraEnv last so that it can override any other values.
-	// With one exception: ENVBUILDER_CACHE_REPO and ENVBUILDER_GIT_URL are required and should not be overridden.
-	// Other values set by the provider may be overridden, but will generate a warning.
-	var diag, ds diag.Diagnostics
-	if !data.ExtraEnv.IsNull() {
-		for key, elem := range data.ExtraEnv.Elements() {
-			switch key {
-			// These are required and should not be overridden.
-			case "ENVBUILDER_CACHE_REPO", "ENVBUILDER_GIT_URL":
-				diag.AddAttributeWarning(path.Root("extra_env"),
-					"Cannot override required environment variable",
-					fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+	// convert extraEnv to a map for ease of use.
+	extraEnv := make(map[string]string)
+	for k, v := range data.ExtraEnv.Elements() {
+		extraEnv[k] = tfValueToString(v)
+	}
+	diags = append(diags, overrideOptionsFromExtraEnv(&opts, extraEnv, overrides)...)
+
+	return opts, diags
+}
+
+func overrideOptionsFromExtraEnv(opts *eboptions.Options, extraEnv map[string]string, overrides map[string]struct{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	// Make a map of the options for easy lookup.
+	optsMap := make(map[string]pflag.Value)
+	for _, opt := range opts.CLI() {
+		optsMap[opt.Env] = opt.Value
+	}
+	for key, val := range extraEnv {
+		switch key {
+
+		// These options may not be overridden.
+		case "ENVBUILDER_CACHE_REPO", "ENVBUILDER_GIT_URL":
+			diags.AddAttributeWarning(path.Root("extra_env"),
+				"Cannot override required environment variable",
+				fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+			)
+			continue
+
+		default:
+			// Check if the option was set on the provider data model and generate a warning if so.
+			if _, overridden := overrides[key]; overridden {
+				diags.AddAttributeWarning(path.Root("extra_env"),
+					"Overriding provider environment variable",
+					fmt.Sprintf("The key %q in extra_env overrides an option set on the provider.", key),
+				)
+			}
+
+			// XXX: workaround for serpent behaviour where calling Set() on a
+			// string slice will append instead of replace: set to empty first.
+			if key == "ENVBUILDER_IGNORE_PATHS" {
+				_ = optsMap[key].Set("")
+			}
```

Comment:
I'm explicitly choosing to overwrite instead of append here as I believe the latter will be suprising and unexpected.

Reply:
Is there a change this could get set to empty but then not replaced?

Reply:
If somehow `ENVBUILDER_IGNORE_PATHS` ended up in `extraEnv` but not in `optsMap`, which would likely result in a nil pointer error on line 457. I could only see that happening if we removed that option.

This whole options situation is horrible and I agree we need a better way to do it. For now, I'm just focusing on getting it working.